src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostOnAutoInsertEndpoint.cs

```diff
@@ -0,0 +1,153 @@
+
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor;
+using Microsoft.AspNetCore.Razor.LanguageServer.Hosting;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost;
+using Microsoft.CodeAnalysis.Razor.AutoInsert;
+using Microsoft.CodeAnalysis.Razor.DocumentMapping;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.Protocol.AutoInsert;
+using Microsoft.CodeAnalysis.Razor.Remote;
+using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Razor.LanguageClient.Cohost;
+using Microsoft.VisualStudio.Razor.LanguageClient.Extensions;
+
+using RazorLSPConstants = Microsoft.VisualStudio.Razor.LanguageClient.RazorLSPConstants;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor.LanguageClient.Cohost;
+
+#pragma warning disable RS0030 // Do not use banned APIs
+[Shared]
+[CohostEndpoint(VSInternalMethods.OnAutoInsertName)]
+[Export(typeof(IDynamicRegistrationProvider))]
+[ExportCohostStatelessLspService(typeof(CohostOnAutoInsertEndpoint))]
+[method: ImportingConstructor]
+#pragma warning restore RS0030 // Do not use banned APIs
+internal class CohostOnAutoInsertEndpoint(
+    IRemoteServiceInvoker remoteServiceInvoker,
+#pragma warning disable RS0030 // Do not use banned APIs
+    [ImportMany] IEnumerable<IOnAutoInsertTriggerCharacterProvider> onAutoInsertTriggerCharacterProviders,
+#pragma warning restore RS0030 // Do not use banned APIs
+    IHtmlDocumentSynchronizer htmlDocumentSynchronizer,
+    LSPRequestInvoker requestInvoker,
+    IRazorDocumentMappingService razorDocumentMappingService,
+    ILoggerFactory loggerFactory)
+    : AbstractRazorCohostDocumentRequestHandler<VSInternalDocumentOnAutoInsertParams, VSInternalDocumentOnAutoInsertResponseItem?>, IDynamicRegistrationProvider
+{
+    private readonly IRemoteServiceInvoker _remoteServiceInvoker = remoteServiceInvoker;
+    private readonly IEnumerable<IOnAutoInsertTriggerCharacterProvider> _onAutoInsertTriggerCharacterProviders = onAutoInsertTriggerCharacterProviders;
+    private readonly IHtmlDocumentSynchronizer _htmlDocumentSynchronizer = htmlDocumentSynchronizer;
+    private readonly LSPRequestInvoker _requestInvoker = requestInvoker;
+    private readonly ILogger _logger = loggerFactory.GetOrCreateLogger<CohostOnAutoInsertEndpoint>();
+    private readonly IRazorDocumentMappingService _razorDocumentMappingService = razorDocumentMappingService;
+
+    protected override bool MutatesSolutionState => false;
+
+    protected override bool RequiresLSPSolution => true;
+
+    public Registration? GetRegistration(VSInternalClientCapabilities clientCapabilities, DocumentFilter[] filter, RazorCohostRequestContext requestContext)
+    {
+        if (clientCapabilities.SupportsVisualStudioExtensions)
```

Comment:
This should be checking for whether dynamic registration is supported for the auto insert capability too. Though its possible there isn't such a thing for VS extensions, I don't know.

Reply:
It's not in VSInternalClientCapabilities. Would it be somewhere else? Is that something that should be added?

Reply:
FWIW I was looking at CohostUriPresentationEndpoint as the model (to copy from ;) ).

Reply:
It's annoying to access, but looks like it's `(clientCapabilities.TextDocument as VSInternalTextDocumentClientCapabilities)?.OnAutoInsert.DynamicRegistration`

Reply:
Thanks, fixed

Reply:
Unfortunately TextDocument is TextDocumentClientCapabilities, not VSInternal variety

![image](https://github.com/user-attachments/assets/f69fb31b-70cc-47a4-a166-466be51e2f05)

so I get

![image](https://github.com/user-attachments/assets/7ae43b51-4d66-4baf-ba88-aebac18d19ea)

I'll revert that check for now until we figure out the proper way to get it working